ArcGISRuntimeSDKQt_CppSamples/UtilityNetwork/ValidateUtilityNetworkTopology/ValidateUtilityNetworkTopology.cpp

```diff
@@ -0,0 +1,671 @@
+        {
+          if (!featureEditResult->error().isEmpty())
+          {
+            m_message = QString("Apply edits completed with error.");
+            emit messageChanged();
+            return;
+          }
+          else
+          {
+            m_message = QString("Apply edits completed successfully.\n"
+                "Click 'Get State' to check the updated network state.");
+            emit messageChanged();
+          }
+        }
+      }
+
+      m_busy = false;
+      emit isBusy();
+    }).onFailed(this, [this]()
+    {
+      m_message = QString("Apply edits failed.");
+      emit messageChanged();
+
+      m_busy = false;
+      emit isBusy();
+    });
+  });
+
+  m_deviceFeatureLayer->clearSelection();
+  m_lineFeatureLayer->clearSelection();
+
+  m_isUpdateVisible = false;
+  emit isUpdateVisible();
+
+  m_validateBtn = true;
+  emit validateBtn();
+
+  m_clearBtn = false;
+  emit clearBtn();
+}
+
+void ValidateUtilityNetworkTopology::onClear()
+{
+  // Clear the selection.
+  m_deviceFeatureLayer->clearSelection();
+  m_lineFeatureLayer->clearSelection();
+
+  m_isUpdateVisible = false;
+  emit isUpdateVisible();
+
+  m_featureToEdit = nullptr;
+
+  m_message = QString("Selection cleared.");
+  emit messageChanged();
+
+  m_clearBtn = false;
+  emit clearBtn();
+}
+
+void ValidateUtilityNetworkTopology::onGetState()
+{
+  m_utilityNetwork = m_mapView->map()->utilityNetworks()->first();
+
+  if(m_utilityNetwork->definition()->capabilities()->isSupportsNetworkState())
+  {
+    m_busy = true;
+    emit isBusy();
+
+    m_message = QString("Getting utility network state...");
+    emit messageChanged();
+
+    m_utilityNetwork->stateAsync().then([this](const QFuture<Esri::ArcGISRuntime::UtilityNetworkState*>& state)
+    {
+      m_getstate = state.result();
+
+      // Validate if dirty areas or errors exist
+      m_validateBtn = m_getstate->hasDirtyAreas();
+      emit validateBtn();
+
+      // Trace if network topology is enabled
+      m_traceBtn = m_getstate->isNetworkTopologyEnabled();
+      emit traceBtn();
+
+      m_message = QString("Utility Network State:\n"
+          "    Has Dirty Areas: " + QString(m_getstate && m_getstate->hasDirtyAreas() ? "true" : "false") + "\n"
+          "    Has Errors: " + QString(m_getstate && m_getstate->hasErrors() ? "true" : "false") + "\n"
+          "    Is Network Topology Enabled: " + QString(m_getstate && m_getstate->isNetworkTopologyEnabled() ? "true" : "false") + "\n");
+
+      if(m_getstate->hasDirtyAreas() || m_getstate->hasErrors())
+      {
+        m_message.append("Click 'Validate' before trace or expect a trace error.");
+      }
+      else
+      {
+        m_message.append("Tap on a feature to edit or click 'Trace' to run a trace.");
+      }
+
+      emit messageChanged();
+    });
+  }
+  m_busy = false;
+  emit isBusy();
+}
+
+void ValidateUtilityNetworkTopology::onValidate()
+{
+  m_utilityNetwork = m_mapView->map()->utilityNetworks()->first();
```

Comment:
Does your logic prevent this from being called before all these are available? If not this will crash. If the service is down, for example, will the user be able to get into this logic?

Reply:
added checking